src/gamemode.c

```diff
@@ -535,6 +535,47 @@ handle_unregister_game_by_pidfd (XdpDbusGameMode *object,
   return G_DBUS_METHOD_INVOCATION_HANDLED;
 }
 
+
+/* properties */
+static void
+update_active_state (GVariant *client_count)
+{
+  gboolean enabled = g_variant_get_int32 (client_count) > 0;
+  xdp_dbus_game_mode_set_active (XDP_DBUS_GAME_MODE (gamemode), enabled);
+}
+
+static void
+update_active_state_from_cache (GDBusProxy *proxy)
+{
+  g_autoptr(GVariant) client_count = NULL;
+
+  client_count = g_dbus_proxy_get_cached_property (proxy, "ClientCount");
+
+  if (client_count != NULL)
+    update_active_state (client_count);
+}
+
+static void
+client_properties_changed (GDBusProxy *proxy,
+                           GVariant *changed_properties,
+                           char **invalidated_properties)
+{
+  const gchar *key;
+  GVariant *value = NULL;
+  g_autoptr(GVariantIter) iter;
+
+  g_variant_get (changed_properties, "a{sv}", &iter);
+  while (g_variant_iter_next (iter, "{&sv}", &key, &value))
+  {
```

Comment:
This whole hash table iteration is unnecessary :slightly_smiling_face: 

The `g-properties-changed` signal conveniently passes the list of invalidated properties for us, that's the `invalidated_properties` array. You can simply check if the `ClientCount` string is in that array. Something like this:

```c
  int n_clients;

  if (!g_strv_contains (invalidated_properties, "ClientCount"))
    return;

  g_variant_lookup (changed_properties, "ClientCount", "i", &n_clients);
  xdp_dbus_game_mode_set_active (XDP_DBUS_GAME_MODE (gamemode), n_clients > 0);
```

Reply:
### invalidated properties
Unless I misunderstand you, I don't think `invalidated_properties` does what you say.

`invalidated_properties` and `changed_properties` are two different lists, [meaning different things](https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-properties).

The invalidated list is basically the other side on the bus being lazy and merely signaling that a property changed but not sending the value. Then the client has to go ask for the value if it actually cares.

Since that's a little annoying when you do care about the value, I added the flag `G_DBUS_PROXY_FLAGS_GET_INVALIDATED_PROPERTIES` so that the invalidated properties list can be ignored (I assume it's always empty) and I can just check the changed properties list instead.

### dictionary lookups

But yeah, I can use the dictionary variant calls instead of visibly looping through the list. Latest update calls `g_variant_lookup_value` instead of looping itself. And then passes the variant to `update_active_state`.